src/shims/env.rs

```diff
@@ -29,42 +30,100 @@ impl<'tcx> EnvVars<'tcx> {
-    fn getenv(&mut self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
-        let this = self.eval_context_mut();
+    fn getenv(&self, name_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_ref();
 
         let name_ptr = this.read_scalar(name_op)?.not_undef()?;
-        let name = this.read_os_str_from_c_str(name_ptr)?;
-        Ok(match this.machine.env_vars.map.get(name) {
-            // The offset is used to strip the "{name}=" part of the string.
+        let name = this.read_os_str_from_target_str(name_ptr)?;
+        Ok(match this.machine.env_vars.map.get(&name) {
             Some(var_ptr) => {
+                // The offset is used to strip the "{name}=" part of the string.
                 Scalar::from(var_ptr.offset(Size::from_bytes(u64::try_from(name.len()).unwrap().checked_add(1).unwrap()), this)?)
             }
             None => Scalar::ptr_null(&*this.tcx),
         })
     }
 
+
+    fn getenvironmentvariablew(
+        &mut self,
+        name_op: OpTy<'tcx, Tag>, // LPCWSTR lpName
+        buf_op: OpTy<'tcx, Tag>, // LPWSTR  lpBuffer
+        size_op: OpTy<'tcx, Tag>, // DWORD   nSize
+    ) -> InterpResult<'tcx, u32> {
+        let this = self.eval_context_mut();
+
+        let name_ptr = this.read_scalar(name_op)?.not_undef()?;
+        let name = this.read_os_str_from_target_str(name_ptr)?;
+        Ok(match this.machine.env_vars.map.get(&name) {
+            Some(var_ptr) => {
+                // The offset is used to strip the "{name}=" part of the string.
+                let var_ptr = Scalar::from(var_ptr.offset(Size::from_bytes((name.len() as u64 + 1) * 2), this)?);
+                let buf_size = this.read_scalar(size_op)?.to_i32()? as u64;
+                let buf_ptr = this.read_scalar(buf_op)?.not_undef()?;
+                let size_u16 = Size::from_bytes(2);
+
+                // The following loop attempts to figure out the length of env_var (`var_size`)
+                let mut var_size = 0u64;
+                loop {
+                    let temp_var_ptr = var_ptr.ptr_offset(Size::from_bytes(var_size * 2), this)?;
+                    let bytes = this.memory.read_bytes(temp_var_ptr, size_u16)?;
+                    var_size += 1;
+                    // encountered 0x0000 terminator
+                    if bytes[0] == 0 && bytes[1] == 0 { break; }
+                }
+
+                let return_val = if var_size > buf_size {
+                    // If lpBuffer is not large enough to hold the data, the return value is the buffer size, in characters,
+                    // required to hold the string and its terminating null character and the contents of lpBuffer are undefined.
+                    var_size
+                } else {
+                    for i in 0..var_size {
+                        this.memory.copy(
+                            this.force_ptr(var_ptr.ptr_offset(Size::from_bytes(i * 2), this)?)?,
```

Comment:
```suggestion
                            this.force_ptr(var_ptr.ptr_offset(Size::from_bytes(i) * 2, this)?)?,
```
This way we get overflow checking. (Same in the next line.)